### PR TITLE
Security Update/Penalty System

### PR DESCRIPTION
**Pull request by E1Code**
Simplified penalty system based on PR by cc:@hackmod  (https://github.com/EthersocialNetwork/go-ethereum/pull/1)

Originally based on the PIRLGuard system developed and deployed by the PIRL Team (https://git.pirl.io/)

Also See:
Horizen(ZenCash) System: https://www.horizen.global/assets/files/A-Penalty-System-for-Delayed-Block-Submission-by-Horizen.pdf
51% Attack Mitigation Penalty System by Horizen - https://www.youtube.com/watch?v=E99wpSZs6iM

### CODE REVIEW

**Comment by masterdubs** [comment]
Be sure to keep the chain sorting ( that was removed in your code ). If you remove it, you can have timing errors and corrupt your chain.

**Comment by hackmod** [comment]
> Be sure to keep the chain sorting ( that was removed in your code ). If you remove it, you can have timing errors and corrupt your chain.

Please see
https://github.com/ethereum/go-ethereum/blob/master/core/blockchain.go#L737-L744

at that points, chain blocks are already sorted.

File: core/penaltysystem.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	delayedBlockWarnLen = 15
+)
+
+var (
+	blockDelayedMeter = metrics.NewRegisteredMeter("chain/block/delayed", nil)
+	blockPenaltyMeter = metrics.NewRegisteredMeter("chain/block/penalty", nil)
+)
+
+// CheckDelayedChain will check possible 51% attack.
+// Penalty System penalize newly inserted blocks.
```
Comment (Primate411) [inline core/penaltysystem.go]:
```suggestion
// Penalty System penalizes newly inserted blocks.
```

File: core/penaltysystem.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+// The amount of penalty depends on the amount of blocks mined by the malicious miner privatly.
```
Comment (Primate411) [inline core/penaltysystem.go]:
```suggestion
// The amount of penalty blocks depends on the amount of blocks mined by the malicious miner privately.
```

File: core/penaltysystem.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+	delayedBlockInfoLen = 3
+	delayedBlockWarnLen = 15
+)
+
+var (
+	blockDelayedMeter = metrics.NewRegisteredMeter("chain/block/delayed", nil)
+	blockPenaltyMeter = metrics.NewRegisteredMeter("chain/block/penalty", nil)
+)
+
+// CheckDelayedChain will check possible 51% attack.
```
Comment (Primate411) [inline core/penaltysystem.go]:
```suggestion
// CheckDelayedChain will check for a possible 51% attack.
```

File: core/penaltysystem.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/metrics"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+var syncing bool
+
+const (
+	description         = "Penalty System (based on the PirlGuard by Pirl Team)"
```
Comment (Primate411) [inline core/penaltysystem.go]:
```suggestion
	description         = "Penalty System (based on PirlGuard by Pirl Team)"
```

File: core/penaltysystem.go (hunk 2 above)
Comment (mkrufky) [reply core/penaltysystem.go]: plural blocks -> singular depend

**Pull request merged**